evm/p2p/peer.py

```diff
@@ -280,48 +338,117 @@ def match_protocols(self, remote_capabilities):
             offset += proto_klass.cmd_length
 
 
+class LESPeer(BasePeer):
+    _les_proto = None
+    _supported_sub_protocols = [LESProtocol]
+
+    def __init__(self, remote, privkey, reader, writer, aes_secret, mac_secret,
+                 egress_mac, ingress_mac, chaindb, network_id):
+        super(LESPeer, self).__init__(
+            remote, privkey, reader, writer, aes_secret, mac_secret,
+            egress_mac, ingress_mac, chaindb, network_id)
+        self._header_fetching_lock = asyncio.Lock()
+
+    @asyncio.coroutine
+    def fetch_headers(self, block_number, reorg_depth=0):
+        """Fetch all headers from our canonical chain head up to block_number.
+
+        If reorg_depth is provided, fetch headers from (current_head - reorg_depth) up to
+        block_number.
+        """
+        with (yield from self._header_fetching_lock):
+            self.logger.info("fetch_headers({}, reorg_depth={}) called".format(
+                block_number, reorg_depth))
+            head_number = self.chaindb.get_canonical_head().block_number
+            head_number -= reorg_depth
+            announced_head_number = block_number
+            while announced_head_number > head_number:
+                request_id = gen_request_id()
+                target_head = min(head_number + MAX_HEADERS_FETCH, announced_head_number)
+                self.les_proto.send_get_block_headers(
+                    target_head, MAX_HEADERS_FETCH, request_id, reverse=True)
+                reply = yield from self.wait_for_reply(request_id)
+                for header in reversed(reply['headers']):
+                    self.chaindb.persist_header_to_db(header)
+                    # FIXME: The reference to the canonical chain head should not be stored in the
+                    # chaindb as that is shared by all peers -- instead, that should be kept on a
+                    # peer-specific db.
+                    self.chaindb.add_block_number_to_hash_lookup(header)
+                    self.chaindb.set_canonical_head(header.hash)
+                    head_number = header.block_number
+                self.logger.info("synced headers up to block {}".format(head_number))
+
+    @property
+    def les_proto(self):
+        """Return the LESProtocol available for this peer.
+
+        The available LESProtocol will be the highest version supported by this class and the
+        remote peer. It is available only after the initial P2P handshake.
+        """
+        assert len(self.enabled_sub_protocols) > 0
+        if self._les_proto is None:
+            for proto in self.enabled_sub_protocols:
+                if proto.name == LESProtocol.name:
+                    self._les_proto = proto
+        return self._les_proto
```

Comment:
This function seems to be missing handling for the case where none of the enabled sub-protocols are the `LESProtocol`.  Is the intention to return `None` in this case?  If so, it seems like a good comment to put in place to make it explicit.

Reply:
Well, this is the LESPeer class, which only supports LESProtocoll and the handshake will fail if there's no matching sub-protocols. Maybe I could add a comment to clarify that?

Reply:
In cases like this when the code path is *unreachable* I like to add a `raise Exception("Invariant")` so that if some future change makes it possible it doesn't fail silently.

Reply:
Makes sense, I've added one